File: python/striped_triangles.py
import random, time, os

r"""
  \        /      \ \ \    / / /
 \ \      / /      \ \      / /
\ \ \    / / /      \        /
"""

DELAY = 0.1
changeAmt = 4

def main():
    density = 0
    while True:
        # -1 because Windows adds newlines if anything
        # is printed in the rightmost column.
        width = os.get_terminal_size()[0] - 1

        # The width of a pair of triangles is 7 characters:
        #   /\ \ \
        #  / /\ \
        # / / /\
        # 123456
        numTrianglePairs = (width - 2) // 6

        # Increase/decrease the density of triangles, until you
        # reach 0 or 100, then reverse the direction of density change:
        density += changeAmt
        if density <= 0 or density >= 100:
            changeAmt *= -1

        # Draw a row that starts with an upright triangle on the left side.
        row1 = ['  ']
        row2 = [' ']
        row3 = []
        for i in range(numTrianglePairs):
            if random.randint(0, 99) < density:
                if random.randint(0, 1):
                    row1.append('\\')
                    row2.append('\\ \\')
                    row3.append('\\ \\ \\')
                else:
                    row1.append('/')
                    row2.append('/ /')
                    row3.append('/ / /')
            else:
                row1.append(' ')
                row2.append('   ')
                row3.append('     ')

            if random.randint(0, 99) < density:
                if random.randint(0, 1):
                    row1.append('\\ \\ \\')
                    row2.append('\\ \\')
                    row3.append('\\')
                else:
                    row1.append('/ / /')
                    row2.append('/ /')
                    row3.append('/')
            else:
                row1.append('     ')
                row2.append('   ')
                row3.append(' ')
        print(''.join(row1)); time.sleep(DELAY)
        print(''.join(row2)); time.sleep(DELAY)
        print(''.join(row3)); time.sleep(DELAY)


        # Draw a row that starts with an upside down triangle on the left side.
        row1 = []
        row2 = [' ']
        row3 = ['  ']
        for i in range(numTrianglePairs):
            if random.randint(0, 99) < density:
                if random.randint(0, 1):
                    row1.append('\\ \\ \\')
                    row2.append('\\ \\')
                    row3.append('\\')
                else:
                    row1.append('/ / /')
                    row2.append('/ /')
                    row3.append('/')
            else:
                row1.append('     ')
                row2.append('   ')
                row3.append(' ')

            if random.randint(0, 99) < density:
                if random.randint(0, 1):
                    row1.append('\\')
                    row2.append('\\ \\')
                    row3.append('\\ \\ \\')
                else:
                    row1.append('/')
                    row2.append('/ /')
                    row3.append('/ / /')
            else:
                row1.append(' ')
                row2.append('   ')
                row3.append('     ')

        print(''.join(row1)); time.sleep(DELAY)
        print(''.join(row2)); time.sleep(DELAY)
        print(''.join(row3)); time.sleep(DELAY)

try:
    main()
except KeyboardInterrupt:
    print('Striped Triangles, by Al Sweigart al@inventwithpython.com')

File: typescript/stripedtriangles.ts
// To get ts types, run: npm i --save-dev @types/readline

namespace StripedTriangles {
    let width: number = 120

    const DELAY: number = 60;
    let CHANGE_AMOUNT: number = 0.06;
    
    async function sleep(milliseconds: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, milliseconds));
    }

    process.on('SIGINT', function() {
        console.log("\nStriped Triangles by Al Sweigart al@inventwithpython.com");
        process.exit();
    });

    function getTerminalWidth(): number {
        if (process.stdout.isTTY) {
            return process.stdout.columns - 1; // -1 is because Windows puts a newline when printing to the last column
        } else {
            return 80;
        }
    }

    async function main(): Promise<void> {
        let density: number = 0.0;

        while (true) {
            width = getTerminalWidth();

            // The width of a pair of triangles is 8 characters:
            //   /\ \ \
            //  / /\ \
            // / / /\
            // 12345678
            let numTrianglePairs: number = Math.floor((width - 2) / 6);

            density += CHANGE_AMOUNT;
            if (density < 0 || density >= 1.0) {
                CHANGE_AMOUNT *= -1;
            }

            // Draw a row that starts with an upright triangle on the left side.
            let row1: string = '  ';
            let row2: string = ' ';
            let row3: string = '';
            for (let i = 0; i < numTrianglePairs; i++) {
                if (Math.random() < density) {
                    if (Math.random() < 0.5) {
                        row1 += '\\';
                        row2 += '\\ \\';
                        row3 += '\\ \\ \\';
                    } else {
                        row1 += '/';
                        row2 += '/ /';
                        row3 += '/ / /';
                    }
                } else {
                    row1 += ' ';
                    row2 += '   ';
                    row3 += '     ';
                }

                if (Math.random() < density) {
                    if (Math.random() < 0.5) {
                        row1 += '\\ \\ \\';
                        row2 += '\\ \\';
                        row3 += '\\';
                    } else {
                        row1 += '/ / /';
                        row2 += '/ /';
                        row3 += '/';
                    }
                } else {
                    row1 += '     ';
                    row2 += '   ';
                    row3 += ' ';
                }
            }
            console.log(row1);
            await sleep(DELAY);
            console.log(row2);
            await sleep(DELAY);
            console.log(row3);
            await sleep(DELAY);
            
            // Draw a row that starts with an upside down triangle on the left side.
            row1 = '';
            row2 = ' ';
            row3 = '  ';
            for (let i = 0; i < numTrianglePairs; i++) {
                if (Math.random() < density) {
                    if (Math.random() < 0.5) {
                        row1 += '\\ \\ \\';
                        row2 += '\\ \\';
                        row3 += '\\';
                    } else {
                        row1 += '/ / /';
                        row2 += '/ /';
                        row3 += '/';
                    }
                } else {
                    row1 += '     ';
                    row2 += '   ';
                    row3 += ' ';
                }

                if (Math.random() < density) {
                    if (Math.random() < 0.5) {
                        row1 += '\\';
                        row2 += '\\ \\';
                        row3 += '\\ \\ \\';
                    } else {
                        row1 += '/';
                        row2 += '/ /';
                        row3 += '/ / /';
                    }
                } else {
                    row1 += ' ';
                    row2 += '   ';
                    row3 += '     ';
                }
            }
            console.log(row1);
            await sleep(DELAY);
            console.log(row2);
            await sleep(DELAY); 
            console.log(row3);
            await sleep(DELAY);
        }
    }

    main();
}